sympy/functions/special/bessel.py

```diff
@@ -223,18 +223,29 @@ def _eval_rewrite_as_jn(self, nu, z, **kwargs):
 
     def _eval_as_leading_term(self, x, logx=None, cdir=0):
         nu, z = self.args
-        arg = z.as_leading_term(x)
-        if x in arg.free_symbols:
+        try:
+            arg = z.as_leading_term(x)
+        except NotImplementedError:
+            return self
+        _, e = arg.as_coeff_exponent(x)
+
+        if e.is_positive:
             return arg**nu/(2**nu*gamma(nu + 1))
-        else:
-            return self.func(nu, z.subs(x, 0))
+        elif e.is_negative:
+            # Refer Abramowitz and Stegun 1965, p. 364 for more information on
+            # asymptotic approximation of besselj function.
+            return sqrt(2)*cos(z - pi*(2*nu + 1)/4)/sqrt(pi*z)
```

Comment:
It seems that this does not hold for $\arg z = \pm\pi$ so the condition `e.is_negative` will not suffice.

Reply:
If the limit of the leading term `c*x**e` is `-oo` then something else should be returned, maybe `self` or `NotImplementedError`. The limit can be determined by setting `x = cdir*t` where `t` is positive. Then the sign of `c*cdir**e*t**e` is given by `c*cdir**e` as `t**e` is positive. The above expression can be returned if `c*cdir**e` is not (real and) negative.

Reply:
But yeah , this look interesting . Will give a thought with respect to other bessel functions too !

Reply:
Okay I did some testing on what you've pointed out and I see that this is good catch . The `-pi` case isn't disturbing but the `pi` case is... for `besselj` and `besseli` . The other two look fine to me currently (still thinking whether they could go wrong somewhere)
The simplest wrong result that can be seen is 
```
>>> limit(besseli(1, 1/x), x, 0, '-')
0
>>> # It's counterpart gives the correct result
>>>
>>> limit(besseli(1, x), x, -oo)
-oo
```
This can surely be fixed by returning self for such cases !

Reply:
Actually [DLMF](https://dlmf.nist.gov/10.30) gives us some help with this 

<img width="829" alt="image" src="https://user-images.githubusercontent.com/87052487/187581401-cbc8f25e-5cbd-430f-b421-13e3cef51e1f.png">

So `besseli(1, 1/x)` is given by 



<img width="292" alt="image" src="https://user-images.githubusercontent.com/87052487/187581497-df02c5ca-accd-4206-b7cc-48588cddd4c6.png">

Now if we only use the current code we have 
```
>>> besseli(1, 1/x).as_leading_term(x, cdir=1)
sqrt(2)*exp(1/x)/(2*sqrt(pi)*sqrt(1/x))
>>>
>>> besseli(1, 1/x).as_leading_term(x, cdir=-1)
sqrt(2)*exp(1/x)/(2*sqrt(pi)*sqrt(1/x))
```

But `sqrt(2)*exp(1/x)/(2*sqrt(pi)*sqrt(1/x))` isn't is sync with the left branch of `besseli(1, 1/x)`

<img width="284" alt="image" src="https://user-images.githubusercontent.com/87052487/187581773-7122bf27-dc13-45ec-bd2b-8d7b0a1e46c7.png">

Hence the wrong result
```
>>> limit(besseli(1, 1/x), x, 0, '-')
0
```
For the left branch we need to use the new expression which DLMF points out through something like this
```
        elif e.is_negative:
            cdir = 1 if cdir == 0 else cdir
            sign = c*cdir**e
            if not sign.is_negative and sign.is_real:
                # Refer Abramowitz and Stegun 1965, p. 377 for more information on
                # asymptotic approximation of besseli function.
                return exp(z)/sqrt(2*pi*z)
            return exp((nu + S.Half)*pi*I)*exp(-z)/sqrt(2*pi*z)

        return super(besseli, self)._eval_as_leading_term(x, logx, cdir)
```
This gives us 
```
>>> besseli(1, 1/x).as_leading_term(x)
sqrt(2)*exp(1/x)/(2*sqrt(pi)*sqrt(1/x))
>>>
>>> besseli(1, 1/x).as_leading_term(x, cdir = -1)
-sqrt(2)*I*exp(-1/x)/(2*sqrt(pi)*sqrt(1/x))
```
This gives us the correct left branch

<img width="317" alt="image" src="https://user-images.githubusercontent.com/87052487/187582252-fff4a542-f6e9-4987-8bce-cdc164f3bafd.png">

Reply:
Hence we would require two different expressions for both branches of `besseli(1, 1/x)`

Reply:
I did good amount of testing for this and my take on each of the bessel functions is as follows
1) `besselk` is clear as both [wolfram](https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/06/02/01/02/0005/) and [DLMF](https://dlmf.nist.gov/10.25#E3) show only 1 expressions throughout for all `arg(z)`, hence no change is required here.

2) For `besseli` we can use the alternative expression provided by [DLMF](https://dlmf.nist.gov/10.30) . Wolfram too provides a similar expression

3) For `besselj` there's surely some improvement to be made as you've pointed out . Though DLMF / Abramowitz and Stegun only provide an expression for `|arg(z)| < pi`, hence I checked out [wolfram](https://functions.wolfram.com/Bessel-TypeFunctions/BesselJ/06/02/03/02/0005/) where the `True` part of the piece seems to be working pretty well .

4) For `bessely`, I currently don't see any errors because the leading term we return has similar branches as `bessely` ( just like `besselk` )

<img width="345" alt="image" src="https://user-images.githubusercontent.com/87052487/187586527-fb1b46b9-a7e6-4abe-9fb6-413cd0451c82.png">

```
>>> bessely(1, 1/x).as_leading_term(x)
sqrt(2)*sqrt(x)*(-3*x*sin(pi/4 - 1/x)/8 - cos(pi/4 - 1/x))/sqrt(pi)
```

<img width="284" alt="image" src="https://user-images.githubusercontent.com/87052487/187586643-4cb2e810-3783-446f-8674-28146df30b33.png">

But yeah similar to  `besselj`, we could try using the `True` part of the piecewise function provided by [wolfram](https://functions.wolfram.com/Bessel-TypeFunctions/BesselY/06/02/03/02/0008/)

5) If we aren't sure for `besselj` and `besseli` we could obviously return `self`. I'll make the commit with `self` for `besselj / bessely` and the expression provided by DLMF for `besseli` and `besselk` need not be changed.

Reply:
> [wolfram](https://functions.wolfram.com/Bessel-TypeFunctions/BesselJ/06/02/03/02/0005/) where the True part of the piece seems to be working pretty well

It may be good for `-oo` but not necessarily for small positive `arg(z)`.
```
In [39]: j1 = sqrt(2)/sqrt(pi*z)*cos(z - pi/4)                                  

In [40]: j2 = I*sqrt(2)/sqrt(pi*z)*cos(z + pi/4)                                

In [43]: besselj(0, 100 + I).n()                                                
Out[43]: 0.0312927864188442 + 0.0906215052004212⋅ⅈ

In [44]: j1.subs(z, 100 + I).n()                                                
Out[44]: 0.0314414864729186 + 0.0905904099222782⋅ⅈ

In [45]: j2.subs(z, 100 + I).n()                                                
Out[45]: 0.024195849675035 + 0.119033842277378⋅ⅈ
```

Reply:
Yeah, I was mainly trying to handle the `arg(z) = pi` case ... my arguments became more centric towards these as you mentioned them.
> It seems that this does not hold for  so the condition e.is_negative will not suffice.

But yeah I get what you're trying to convey, currently I don't think we have good reference for addressing this, hence as you've mentioned above 
> something else should be returned, maybe self or NotImplementedError

could be returned. I personally like `self` because I've noticed `heuristics` work well with `bessel` functions and hence there are more chances of getting the value we want as output .

Reply:
The true piece of wolfram might also work for - oo.

Reply:
Yess they do .All expressions I have referenced above were done keeping `-oo` in mind .

Reply:
Yeah , strictly speaking with respect to `oo/-oo` , I would prefer the following results 
```
>>> besselj(1, 1/x).as_leading_term(x, cdir = 1)
-sqrt(2)*sin(pi/4 - 1/x)/(sqrt(pi)*sqrt(1/x))
>>> besselj(1, 1/x).as_leading_term(x, cdir = -1)
sqrt(2)*I*sqrt(x)*sin(pi/4 + 1/x)/sqrt(pi)
>>> bessely(1, 1/x).as_leading_term(x, cdir = 1)
sqrt(2)*sqrt(x)*(-3*x*sin(pi/4 - 1/x)/8 - cos(pi/4 - 1/x))/sqrt(pi)
>>> bessely(1, 1/x).as_leading_term(x, cdir = -1)
-sqrt(2)*sqrt(x)*(-exp(3*I*pi/4 + I/x) + cos(pi/4 - 1/x))/sqrt(pi)
>>> besseli(1, 1/x).as_leading_term(x, cdir = 1)
sqrt(2)*exp(1/x)/(2*sqrt(pi)*sqrt(1/x))
>>> besseli(1, 1/x).as_leading_term(x, cdir = -1)
-sqrt(2)*I*exp(-1/x)/(2*sqrt(pi)*sqrt(1/x))
```
Over something like
```
>>> besselj(1, 1/x).as_leading_term(x, cdir = 1)
-sqrt(2)*sin(pi/4 - 1/x)/(sqrt(pi)*sqrt(1/x))
>>> besselj(1, 1/x).as_leading_term(x, cdir = -1)
besselj(1, 1/x)
>>> bessely(1, 1/x).as_leading_term(x, cdir = 1)
sqrt(2)*sqrt(x)*(-3*x*sin(pi/4 - 1/x)/8 - cos(pi/4 - 1/x))/sqrt(pi)
>>> bessely(1, 1/x).as_leading_term(x, cdir = -1)
bessely(1, 1/x)
>>> besseli(1, 1/x).as_leading_term(x, cdir = 1)
sqrt(2)*exp(1/x)/(2*sqrt(pi)*sqrt(1/x))
>>> besseli(1, 1/x).as_leading_term(x, cdir = -1)
besseli(1, 1/x)
```
I am fully sure about `Besseli` , for others we are using the true parts from wolfram and I see that they work quite well so maybe we could go with this .
```
--- a/sympy/functions/special/bessel.py
+++ b/sympy/functions/special/bessel.py
@@ -238,7 +238,7 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
                 # Refer Abramowitz and Stegun 1965, p. 364 for more information on
                 # asymptotic approximation of besselj function.
                 return sqrt(2)*cos(z - pi*(2*nu + 1)/4)/sqrt(pi*z)
-            return self
+            return sqrt(2)*I*sqrt(1/z)*exp(I*pi*nu)*cos(z + pi*(2*nu + 1)/4)/sqrt(pi)

         return super(besselj, self)._eval_as_leading_term(x, logx, cdir)

@@ -376,7 +376,7 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
                 # Refer Abramowitz and Stegun 1965, p. 364 for more information on
                 # asymptotic approximation of bessely function.
                 return sqrt(2)*(-sin(pi*nu/2 - z + pi/4) + 3*cos(pi*nu/2 - z + pi/4)/(8*z))*sqrt(1/z)/sqrt(pi)
-            return self
+            return -sqrt(2)*(exp(I*pi*nu/2 + I*z + I*pi/4)*cos(pi*nu) + cos(-pi*nu/2 + z + pi/4))*sqrt(1/z)/sqrt(pi)

         return super(bessely, self)._eval_as_leading_term(x, logx, cdir)

@@ -551,7 +551,7 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
                 # Refer Abramowitz and Stegun 1965, p. 377 for more information on
                 # asymptotic approximation of besseli function.
                 return exp(z)/sqrt(2*pi*z)
-            return self
+            return exp((nu + S.Half)*pi*I)*exp(-z)/sqrt(2*pi*z)

         return super(besseli, self)._eval_as_leading_term(x, logx, cdir)
```

Reply:
The true part should work as we already cover `|arg(z)| < pi` through our current implementation and as arg function can take values between `− 𝜋 < 𝜃 ≤ 𝜋` , we might only have to take care of `arg(z) == pi` , which is done correctly from the true part .

Reply:
Can we try something like the following for `besselj` and `bessely`
```
--- a/sympy/functions/special/bessel.py
+++ b/sympy/functions/special/bessel.py
@@ -222,6 +222,7 @@ def _eval_rewrite_as_jn(self, nu, z, **kwargs):
         return sqrt(2*z/pi)*jn(nu - S.Half, self.argument)

     def _eval_as_leading_term(self, x, logx=None, cdir=0):
+        from sympy.functions.elementary.complexes import arg as _arg
         nu, z = self.args
         try:
             arg = z.as_leading_term(x)
@@ -232,13 +233,11 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
         if e.is_positive:
             return arg**nu/(2**nu*gamma(nu + 1))
         elif e.is_negative:
-            cdir = 1 if cdir == 0 else cdir
-            sign = c*cdir**e
-            if not sign.is_negative and sign.is_real:
-                # Refer Abramowitz and Stegun 1965, p. 364 for more information on
-                # asymptotic approximation of besselj function.
-                return sqrt(2)*cos(z - pi*(2*nu + 1)/4)/sqrt(pi*z)
-            return self
+            if _arg(cdir) == S.Pi:
+                return sqrt(2)*I*sqrt(1/z)*exp(I*pi*nu)*cos(z + pi*(2*nu + 1)/4)/sqrt(pi)
+            # Refer Abramowitz and Stegun 1965, p. 364 for more information on
+            # asymptotic approximation of besselj function.
+            return sqrt(2)*cos(z - pi*(2*nu + 1)/4)/sqrt(pi*z)

         return super(besselj, self)._eval_as_leading_term(x, logx, cdir)

@@ -356,6 +355,7 @@ def _eval_rewrite_as_yn(self, nu, z, **kwargs):
         return sqrt(2*z/pi) * yn(nu - S.Half, self.argument)

     def _eval_as_leading_term(self, x, logx=None, cdir=0):
+        from sympy.functions.elementary.complexes import arg as _arg
         nu, z = self.args
         try:
             arg = z.as_leading_term(x)
@@ -370,13 +370,11 @@ def _eval_as_leading_term(self, x, logx=None, cdir=0):
             arg = Add(*[term_one, term_two, term_three]).as_leading_term(x, logx=logx)
             return arg
         elif e.is_negative:
-            cdir = 1 if cdir == 0 else cdir
-            sign = c*cdir**e
-            if not sign.is_negative and sign.is_real:
-                # Refer Abramowitz and Stegun 1965, p. 364 for more information on
-                # asymptotic approximation of bessely function.
-                return sqrt(2)*(-sin(pi*nu/2 - z + pi/4) + 3*cos(pi*nu/2 - z + pi/4)/(8*z))*sqrt(1/z)/sqrt(pi)
-            return self
+            if _arg(cdir) == S.Pi:
+                return -sqrt(2)*(exp(I*pi*nu/2 + I*z + I*pi/4)*cos(pi*nu) + cos(-pi*nu/2 + z + pi/4))*sqrt(1/z)/sqrt(pi)
+            # Refer Abramowitz and Stegun 1965, p. 364 for more information on
+            # asymptotic approximation of besselj function.
+            return sqrt(2)*(-sin(pi*nu/2 - z + pi/4) + 3*cos(pi*nu/2 - z + pi/4)/(8*z))*sqrt(1/z)/sqrt(pi)

         return super(bessely, self)._eval_as_leading_term(x, logx, cdir)
```

Reply:
What do you think of the above pasted comments ? @jksuom